### PR TITLE
Add string constant provider

### PR DESCRIPTION
**Pull request by bhagenbourger**
Partially solves #20

### CODE REVIEW

**Comment by bhagenbourger** [comment]
I only manage string type for the moment for string provider. I will do another PR for other types later.  
Is this what you expected for constant provider?

File: tests/parquet_all_options.yaml [hunk 1]
```diff
@@ -42,6 +42,14 @@ columns:
...
     min: -100
     max: 100
+  
+  - name: constant_string
+    provider: Constant.string
+    data: 
+      - value: 42
+      - value: my_second_constant
+        weight: 8
+      - value: my_third_constant
```
Comment (vianneybacoup) [inline tests/parquet_all_options.yaml:52]: Can you add all the different cases of setting data here ?
There is a test that uses both all_options.yaml file

File: src/providers/provider.rs [hunk 2]
```diff
@@ -79,38 +82,21 @@ mod tests {
...
-    #[test]
-    fn given_random_should_return_provider() {
-        let provider_name = "random.string.alphanumeric";
-        let yaml_str = format!("name: name{}provider: {}", '\n', provider_name);
-        let column = &YamlLoader::load_from_str(yaml_str.as_str()).unwrap()[0];
-
-        match ProviderBuilder::get_corresponding_provider(provider_name, column) {
-            Ok(_) => (),
-            _ => panic!(),
+    fn given_valid_provider_should_return_provider() {
```
Comment (vianneybacoup) [inline src/providers/provider.rs:85]: Way cleaner, thanks

**Review by vianneybacoup — COMMENTED** [review]
@bhagenbourger I imagined only the part with only one value, so you went waayy further ! Thank you for this ! 👍 
I put some comments, feel free to question them if you feel they are not relevant.
I also let you rebase, there are some updates to add as well.

**Comment by bhagenbourger** [comment]
Thank you for your feedback @vianneybacoup 
I think about your advice and I propose to split what I did into three providers:
 - ConstantStringProvider : a provider which accepts a single value and returns this value
 - ListStringProvider : a provider which accepts a list of values and returns a value randomly chosen
 - WeightedListStringProvider : a provider which accepts a weighted list of values and returns a value weighted randomly chosen

What do you think about that?

**Comment by vianneybacoup** [comment]
> Thank you for your feedback @vianneybacoup I think about your advice and I propose to split what I did into three providers:
> 
> * ConstantStringProvider : a provider which accepts a single value and returns this value
> * ListStringProvider : a provider which accepts a list of values and returns a value randomly chosen
> * WeightedListStringProvider : a provider which accepts a weighted list of values and returns a value weighted randomly chosen
> 
> What do you think about that?

That's perfect.
This should be purely on dev side. For the user, it will access those three providers with Constant.string (like what you did already), and the new_from_yaml function will return the correct one.

File: src/providers/parameters/wstring.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            Yaml::Hash(h) => extrat_hash(column_name, h, default_value),
+            Yaml::String(s) => WStringParameter::new_from_str(s.as_str()),
+            Yaml::Integer(i) => WStringParameter::new_from_i64(i),
+            Yaml::Real(r) => WStringParameter::new_from_str(r),
+            _ => WStringParameter::new_from_str(default_value),
+        })
+        .collect::<Vec<_>>()
+}
+
+fn extrat_hash(
```
[the remark this answers is not in the document]
Comment (bhagenbourger) [reply src/providers/parameters/wstring.rs]: Sorry didn't see it, fixed.

File: Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ once_cell = "1.8"
 parquet = "50.0.0"
 rayon = "1.8.0"
 yaml-rust = "0.4"
+linked-hash-map = "0.5.6"
```
[the remark this answers is not in the document]
Comment (bhagenbourger) [reply Cargo.toml]: I didn't pay attention for alphabetical order, fixed

File: src/providers/constant/string.rs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+use crate::providers::provider::{Provider, Value};
+
+const DEFAULT_CONSTANT: &str = "constant";
+
+#[derive(Clone)]
+pub struct StringProvider {
+    data: Vec<WStringParameter>,
+}
+
+impl StringProvider {
```
[the remark this answers is not in the document]
Comment (bhagenbourger) [reply src/providers/constant/string.rs]: Renamed. I created 3 providers like discussed.

File: src/providers/constant/string.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    fn weighted_random(weighted_strings: Vec<WStringParameter>) -> String {
+        let max: u32 = weighted_strings.iter().map(|w| w.weight).sum::<u32>();
```
[the remark this answers is not in the document]
Comment (bhagenbourger) [reply src/providers/constant/string.rs]: I asked myself when I wrote this code and I decided that the Vec has few values but definitively this code doesn't scale. I fixed it.

File: src/providers/constant/string.rs [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    data: Vec<String>,
+}
+
+impl Provider for ListStringProvider {
+    fn value(&self, _: u32) -> Value {
+        let index = fastrand::usize(..self.data.len());
+        Value::String(self.data[index].to_string())
+    }
+
+    fn corrupted_value(&self, _: u32) -> Value {
```
Comment (vianneybacoup) [inline src/providers/constant/string.rs:35]: missing tests for all three corrupted_value functions

**Review by vianneybacoup — COMMENTED** [review]
Thanks for all the work !
Nothing too serious to change, we can push this right after

**Review by vianneybacoup — CHANGES_REQUESTED** [review]
I forgot that sorry

**Comment by bhagenbourger** [comment]
No worries, I didn't have time to fix that sooner. I just amended my commit with the fixes.

**Review by vianneybacoup — APPROVED** [review]
Didn't meant to be pushy, sorry about that.
We obviously love the work you made on Fakelake :)

Approved, thanks again

**Pull request merged**